DigitalLearningSolutions.Web/Services/EmailVerificationService.cs

```diff
@@ -0,0 +1,115 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.Models;
+    using DigitalLearningSolutions.Data.Models.Email;
+    using DigitalLearningSolutions.Data.Models.User;
+    using DigitalLearningSolutions.Data.Utilities;
+    using MimeKit;
+
+    public interface IEmailVerificationService
+    {
+        bool AccountEmailIsVerifiedForUser(int userId, string email);
+
+        void SendVerificationEmails(
+            UserAccount userAccount,
+            IEnumerable<PossibleEmailUpdate> unverifiedEmails,
+            string baseUrl
+        );
+    }
+
+    public class EmailVerificationService : IEmailVerificationService
+    {
+        private readonly IEmailVerificationDataService emailVerificationDataService;
+        private readonly IEmailService emailService;
+        private readonly IClockUtility clockUtility;
+
+        public EmailVerificationService(
+            IEmailVerificationDataService emailVerificationDataService,
+            IEmailService emailService,
+            IClockUtility clockUtility
+        )
+        {
+            this.emailVerificationDataService = emailVerificationDataService;
+            this.emailService = emailService;
+            this.clockUtility = clockUtility;
+        }
+
+        public bool AccountEmailIsVerifiedForUser(int userId, string email)
+        {
+            return emailVerificationDataService.AccountEmailIsVerifiedForUser(userId, email);
+        }
+
+        public void SendVerificationEmails(
+            UserAccount userAccount,
+            IEnumerable<PossibleEmailUpdate> unverifiedEmails,
+            string baseUrl
+        )
+        {
+            foreach (var emailGroup in unverifiedEmails.GroupBy(emailUpdate => emailUpdate.NewEmail))
+            {
+                var hash = Guid.NewGuid().ToString();
+                var hashId = emailVerificationDataService.CreateEmailVerificationHash(hash, clockUtility.UtcNow);
+
+                foreach (var emailUpdate in emailGroup)
+                {
+                    UpdateEmailVerificationHashId(userAccount.Id, emailUpdate.CentreId, hashId);
+                }
+
+                emailService.SendEmail(
+                    GenerateVerificationEmail(userAccount, hash, emailGroup.Key!, baseUrl)
+                );
+            }
+        }
+
+        private void UpdateEmailVerificationHashId(int userId, int? centreId, int hashId)
+        {
+            if (centreId == null)
+            {
+                emailVerificationDataService.UpdateEmailVerificationHashIdForPrimaryEmail(userId, hashId);
+            }
+            else
+            {
+                emailVerificationDataService.UpdateEmailVerificationHashIdForCentreEmail(
+                    userId,
+                    centreId.Value,
+                    hashId
+                );
+            }

```

Comment:
I have a suggestion in another comment which should allow you to simplify this method to:
```
emailVerificationDataService.UpdateEmailVerificationHashIdForPrimaryEmail(userId, emailAddress, hashId);
emailVerificationDataService.UpdateEmailVerificationHashIdForCentreEmails(userId, emailAddress, hashId);
```

Note: you shouldn't need to update centre emails in a loop - it is possible to update the verification hash ids for all user/email combinations at once

Reply:
As a side-effect of these changes, `PossibleEmailUpdate` no longer needs to have a `CentreId` property